File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: list[int], target: int) -> int:
        nums.sort()  # Sort the list first
        closest = float('inf')  # Best sum found so far
        n = len(nums)

        for i in range(n - 2):  # Loop through all triplets
            left = i + 1
            right = n - 1

            while left < right:
                total = nums[i] + nums[left] + nums[right]

                # Update closest if this sum is better
                if abs(target - total) < abs(target - closest):
                    closest = total

                # Move pointers
                if total < target:
                    left += 1
                elif total > target:
                    right -= 1
                else:
                    return total  # Exact match, best possible

        return closest
        

File: 0016-3sum-closest/0016-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int t) {
        int closest = 1e8;
        int n = nums.size();
        std::sort(nums.begin(), nums.end());
        for(int i=0; i<n-1; i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            int l = i+1;
            int r = n-1;
            while(l<r){
                int sum = nums[i]+nums[l]+nums[r];
                if(abs(closest-t)>abs(t-sum))closest = sum;
                if(sum>t) r--;
                else if(sum<t) l++;
                else return t;
            }
        }
        return closest;
    }
};